src/macaron/slsa_analyzer/checks/provenance_available_check.py

```diff
@@ -67,6 +75,246 @@ def __init__(self) -> None:
+                try:
+                    data_entry.provenances[prov_asset.url] = load_provenance(
+                        provenance_filepath,
+                    )
+                except ProvenanceLoadError as error:
+                    logger.error("Error while loading provenance: %s", error)
+                    continue
+
+    def find_provenance_assets_on_ci_services(
+        self,
+        repo_full_name: str,
+        ci_info_entries: list[CIInfo],
+        provenance_extensions: list[str],
+    ) -> list[ProvenanceAvailableTable]:
+        """Find provenance assets on CI services.
+
+        Note that we stop going through the CI services once we encounter a CI service
+        that does host provenance assets.
+
+        This method also loads the provenance payloads into the ``CIInfo`` object where
+        the provenance assets are found.
+
+        Parameters
+        ----------
+        repo_full_name: str
+            The full name of the repo, in the format of ``owner/repo_name``.
+        package_registry_data_entries : list[PackageRegistryData]
+            A list of package registry data entries.
+        provenance_extensions : list[str]
+            A list of provenance extensions. Assets with these extensions are assumed
+            to be provenances.
+
+        Returns
+        -------
+        list[ProvenanceAvailableTable]
+            A list of rows to write into the provenance available table. In case no
+            provenance assets found, return an empty list.
+        """
+        for ci_info in ci_info_entries:
+            ci_service = ci_info["service"]
+
+            if isinstance(ci_service, NoneCIService):
+                continue
+
+            # Only get the latest release.
+            release = ci_service.api_client.get_latest_release(repo_full_name)
+            if not release:
+                logger.info("Did not find any release on %s.", ci_service.name)
+                continue
+
+            # Store the release data for other checks.
+            ci_info["latest_release"] = release
+
+            # Get the provenance assets.
+            for prov_ext in provenance_extensions:
+                provenance_assets = ci_service.api_client.get_assets(release, ext=prov_ext)
+                if not provenance_assets:
+                    continue
+
+                logger.info("Found the following provenance assets:")
+                for provenance_asset in provenance_assets:
+                    logger.info("* %s", provenance_asset["url"])
+
+                # Store the provenance assets for other checks.
+                ci_info["provenance_assets"].extend(provenance_assets)
+
+                # Download the provenance assets and load the provenance payloads.
+                self.download_provenances_from_github_actions_ci_service(
+                    ci_info,
+                )
+
+                return [
+                    ProvenanceAvailableTable(
```

Comment:
This line adds a row to `ProvenanceAvailableTable`, which should instead be added in the `run_check` method after deciding if the check is passed or not. Please just return the necessary information here.

Reply:
I wonder if by only create an instance of `ProvenanceAvailableTable` class, the data is automatically added into the database using SQLAlchemy?

Reply:
I think the result rows are not added to the database until [here](https://github.com/oracle/macaron/blob/fcc9b0a4017862ad0cf69e3571913df4d44f8418/src/macaron/slsa_analyzer/database_store.py#L62), isn't it? In these `find_provenance_assets_*` methods, I just "prepare" the rows if there are provenances found and let the `run_check` method decides whether to write these rows into `check_result["result_tables"]`, which will then will be written into the database later on after the check finishes.
A good thing about making these `find_provenance_assets_*` methods return table rows is that it makes use of a readily-defined datatype of table rows. Previously in the implementation of this check, after provenances are discovered on CI services, they are stored inside an intermediate dictionary which is then converted to a list of table rows type right away, and I feel like this dictionary did not serve any obvious purpose. I can keep using this pattern and make these two methods `find_provenance_assets_*` return 2 lists of dictionaries instead of 2 lists of table rows, but again I'm not sure I see a reason for using dictionaries over an already-defined class.

Reply:
When `ProvenanceAvailableTable` is instantiated, a [transient](https://docs.sqlalchemy.org/en/20/orm/session_state_management.html#quickie-intro-to-object-states) instance is created by SQLAlchemy, which will be persisted later when it is added to the session. Creating such transient instances should be done with care and it's best to be done where the check status is determined to be passed in our case to avoid accidental addition of rows to the table. 

I do agree though that the intermediate dictionary can be removed if not necessary. But it's best to address that as part of the new data model [PR](https://github.com/oracle/macaron/pull/305).

Reply:
I think creating transient objects is OK as long as I make sure not to add them to the database.
Anyhow, in the current version, I have changed the return type of the two `find_provenance_assets_*` methods.